packages/calypso-router/src/index.js

```diff
@@ -0,0 +1,1207 @@
+ * @param {string} path
+ * @param {Object=} state
+ * @param {boolean=} dispatch
+ * @param {boolean=} push
+ * @returns {!Context}
+ */
+
+Page.prototype.show = function ( path, state, dispatch, push ) {
+	const ctx = new Context( path, state, this );
+	const prev = this.prevContext;
+	this.prevContext = ctx;
+	this.current = ctx.path;
+	if ( false !== dispatch ) {
+		this.dispatch( ctx, prev );
+	}
+	if ( false !== ctx.handled && false !== push ) {
+		ctx.pushState();
+	}
+	return ctx;
+};
+
+/**
+ * Goes back in the history
+ * Back should always let the current route push state and then go back.
+ * @param {string} path - fallback path to go back if no more history exists, if undefined defaults to page.base
+ * @param {Object=} state
+ */
+
+Page.prototype.back = function ( path, state ) {
+	if ( this.len > 0 ) {
+		const window = this._window;
+		// this may need more testing to see if all browsers
+		// wait for the next tick to go back in history
+		hasHistory && window.history.back();
+		this.len--;
+	} else if ( path ) {
+		setTimeout( () => {
+			this.show( path, state );
+		}, 0 );
+	} else {
+		setTimeout( () => {
+			this.show( this._getBase(), state );
+		}, 0 );
+	}
+};
+
+/**
+ * Register route to redirect from one path to other
+ * or just redirect to another route
+ * @param {string} from - if param 'to' is undefined redirects to 'from'
+ * @param {string=} to
+ */
+Page.prototype.redirect = function ( from, to ) {
+	// Define route from a path to another
+	if ( 'string' === typeof from && 'string' === typeof to ) {
+		page.call( this, from, () => {
+			setTimeout( () => {
+				this.replace( /** @type {!string} */ ( to ) );
+			}, 0 );
+		} );
+	}
+
+	// Wait for the push state and replace it with another
+	if ( 'string' === typeof from && 'undefined' === typeof to ) {
+		setTimeout( () => {
+			this.replace( from );
+		}, 0 );
+	}
+};
+
+/**
+ * Replace `path` with optional `state` object.
+ *
+ * @param {string} path
+ * @param {Object=} state
+ * @param {boolean=} init
+ * @param {boolean=} dispatch
+ * @returns {!Context}
+ */
+
+Page.prototype.replace = function ( path, state, init, dispatch ) {
+	const ctx = new Context( path, state, this );
+	const prev = this.prevContext;
+	this.prevContext = ctx;
+	this.current = ctx.path;
+	ctx.init = init;
+	ctx.save(); // save before dispatching, which may redirect
+	if ( false !== dispatch ) {
+		this.dispatch( ctx, prev );
+	}
+	return ctx;
+};
+
+/**
+ * Dispatch the given `ctx`.
+ * @param {Context} ctx
+ */
+Page.prototype.dispatch = function ( ctx, prev ) {
+	let i = 0;
+	let j = 0;
+
+	const nextEnter = () => {
+		if ( ctx.path !== this.current ) {
+			ctx.handled = false;
+			return;
+		}
+		let fn;
+		while ( 1 ) {
+			fn = this.callbacks[ i++ ];
+			if ( ! fn ) {
+				return unhandled.call( this, ctx );
+			}
+			if ( fn.match( ctx ) ) {
+				break;
+			}
+		}
+		fn( ctx, nextEnter );
+	};
+
+	const nextExit = () => {
+		let fn;
+		while ( 1 ) {
```

Comment:
Any chance any of these might result in an infinite loop?

Reply:
I think it's impossible. The first two lines in the loop body are:
```js
fn = this.callbacks[ i++ ];
if ( ! fn ) {
  return;
}
```
The index variable `i` is unconditionally incremented on each iteration, and we break out of the loop when we reach the end of the `this.callbacks` array. The array would have to be infinite.